core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+
+        let mut applied_snapshot_status = storage
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await?;
+
+        if !storage.blocks_dal().is_genesis_needed().await? && applied_snapshot_status.is_none() {
+            return Err(SnapshotsApplierError::Canceled(
+                "This node has already been initialized without a snapshot".to_string(),
+            ));
+        }
+
+        if let Some(applied_snapshot_status) = applied_snapshot_status {
+            if !applied_snapshot_status
+                .storage_logs_chunks_processed
+                .contains(&false)
+            {
+                return Err(SnapshotsApplierError::Canceled(
+                    "This node has already been initialized from a snapshot".to_string(),
+                ));
+            }
+
+            let recovery = Self {
+                connection_pool,
+                blob_store,
+                applied_snapshot_status,
+            };
+
+            recovery.recover_storage_logs().await?;
+        } else {
+            applied_snapshot_status =
+                Some(SnapshotsApplier::create_fresh_recovery_status(main_node_client).await?);
+
+            let mut recovery = Self {
+                connection_pool,
+                blob_store,
+                applied_snapshot_status: applied_snapshot_status.unwrap(),
+            };
+
+            let latency = latency.observe();
+            tracing::info!("Initialized snapshots applier in {latency:?}");
+
+            METRICS.storage_logs_chunks_count.set(
+                recovery
+                    .applied_snapshot_status
+                    .storage_logs_chunks_processed
+                    .len(),
+            );
+
+            METRICS.storage_logs_chunks_left_to_process.set(
+                recovery
+                    .applied_snapshot_status
+                    .storage_logs_chunks_processed
+                    .iter()
+                    .filter(|x| !(**x))
+                    .count(),
+            );
+
+            recovery.recover_factory_deps(&mut storage).await?;
+
+            storage
+                .snapshot_recovery_dal()
+                .insert_initial_recovery_status(&recovery.applied_snapshot_status)
+                .await?;
+
+            storage.commit().await?;
+
+            recovery.recover_storage_logs().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn create_fresh_recovery_status(
+        main_node_client: &'a dyn SnapshotsApplierMainNodeClient,
+    ) -> anyhow::Result<SnapshotRecoveryStatus> {
+        let snapshot_response = main_node_client.fetch_newest_snapshot().await?;
+
+        let snapshot = snapshot_response.ok_or(SnapshotsApplierError::Canceled(
+            "Main node does not have any ready snapshots, skipping initialization from snapshot!"
+                .to_string(),
+        )).context("No snapshots were found in main node")?;
```

Comment:
About the first point, I think It's an extreme edge-case that and I prefer not to add this to the config, imo it will only confuse the users.
About the second point, I didn't yet catch what you mean, you suggest having an error attached to canceled instead of a string? If so, why is it better?

Reply:
> imo it will only confuse the users.

Fair enough. Just to provide more context: I think snapshot recovery will be an experimental, opt-in feature initially; at least, I'd consider it safe to position it as such. So if we don't provide configurability in this scenario, a safe option would be to halt, so that a user can either fix their setup (network connectivity, GCS bucket options etc.) and retry, or switch off snapshot recovery entirely and sync using other means.

> you suggest removing that String from SnappshotsApplierError::Canceled and attaching message via context?

No, I don't propose that. My line or reasoning is as follows. `snapshot_response.ok_or(_)` returns `Result<_, SnapshotsApplierError>`. Then, you can `anyhow::Context::context` on this `Result`, wrapping an error in  `anyhow::Error`. The question is: why not just create an `anyhow::Error` directly using `snapshot_response.context(_)`? My guess that I've mentioned above is using `root_cause()` of the error when it is handled, which doesn't look like a particularly good idea.

Reply:
Corrected. About the config, I think it's best to return to this conversation on resiliency PR, but also I misunderstood earlier what you want to encode in node config, everything what you have written now makes perfect sense